src/feature.js

```diff
@@ -1,144 +1,184 @@
-   * - set image, text, click handler (telemetry)
-   */
-  constructor(variation) {
-    console.log(
-      "Initializing BrowserActionButtonChoiceFeature:",
-      variation.name,
-    );
-    this.timesClickedInSession = 0;
-
-    // modify BrowserAction (button) ui for this particular {variation}
-    console.log("path:", `icons/${variation.name}.svg`);
-    // TODO: Running into an error "values is undefined" here
-    browser.browserAction.setIcon({ path: Feature.iconPath(variation) });
-    browser.browserAction.setTitle({ title: variation.name });
-    browser.browserAction.onClicked.addListener(() => this.handleButtonClick());
-    console.log("initialized");
+  async handleActivated(activeInfo) {
+    const tab = await browser.tabs.get(activeInfo.tabId);
+    this.activatedTabId = tab.id;
+    this.maybeUpdateSettingListener(tab.id, tab.url);
   }
 
-  /** handleButtonClick
-   *
-   * - instrument browserAction button clicks
-   * - change label
-   */
-  handleButtonClick() {
-    console.log("handleButtonClick");
-    // note: doesn't persist across a session, unless you use localStorage or similar.
-    this.timesClickedInSession += 1;
-    console.log("got a click", this.timesClickedInSession);
-    browser.browserAction.setBadgeText({
-      text: this.timesClickedInSession.toString(),
-    });
+  async handleUpdated(tabId, changeInfo, tabInfo) {
+    if (!changeInfo.url) {
+      return;
+    }
 
-    // telemetry: FIRST CLICK
-    if (this.timesClickedInSession === 1) {
-      browser.study.sendTelemetry({ event: "button-first-click-in-session" });
+    const url = changeInfo.url;
+    console.log(`tab update : TabId: ${tabId}, URL changed to ${url}`);
+    this.maybeUpdateSettingListener(tabId, url);
+    if (!isSupportURLProtocol(url)) {
+      return;
     }
 
-    // telemetry EVERY CLICK
-    browser.study.sendTelemetry({
-      event: "button-click",
-      timesClickedInSession: "" + this.timesClickedInSession,
-    });
+    const pageId = await getBaseDomainHash(url);
+    this.feature.update("visitPage", pageId);
+  }
+
+  handleRemoved(tabId, removeInfo) {
+    if (this.activatedTabId === tabId &&
+        browser.autoplay.autoplaySettingChanged.hasListener(this.settingListener)) {
+      browser.autoplay.autoplaySettingChanged.removeListener(this.settingListener);
+    }
+  }
+}
 
-    // webExtension-initiated ending for "used-often"
-    //
-    // - 3 timesClickedInSession in a session ends the study.
-    // - see `../Config.jsm` for what happens during this ending.
-    if (this.timesClickedInSession >= 3) {
-      browser.study.endStudy("used-often");
+class Feature {
+  constructor() {}
+
+  configure(studyInfo) {
+    const feature = this;
+    const { variation, isFirstRun } = studyInfo;
+    console.log(studyInfo);
+
+    this.tabsMonitor = new TabsMonitor();
+    this.tabsMonitor.configure(this);
+    const sendPingIntervalMS = 1 * 24 * 60 * 60 * 1000;
+    this.sendPingsScheduler(sendPingIntervalMS);
```

Comment:
What I also wanted to remark: There are a bunch of scenarios in which users would not send pings when just scheduling them via setInterval. This could skew your results in favor of users who rarely restart their browser. Sending the telemetry on experiment end helps a bit but you should still consider storing the timeout value (or using shield telemetry which AFAIK aggregates telemetry probes before sending).

I won't require it, however, just wanted to note this.

Reply:
@johannhof Sorry, I don't get it. Why users who rarely restart their browser would affect the the ping sending by `setInterval()`? 

Here, I want to send the ping when one of following conditions happens (1) close browser (2) not close browser, but time passes a day. Therefore, I've add the `sendTelemetryPings()` in the `AutoplayAPI::onShutdown()`.

Thanks!

Reply:
Ok, yeah, sounds like I was wrong about this :)